### ISSUE TITLE
GUI on map functions

### ISSUE THREAD

**Issue by CAPCHIK** [brianzinn/react-babylonjs#75]
GUI elements in _adtFullscreenUi_ are created correctly only on first render, if initial array was changed, gui elements do not change.
So, I have this initial state
```ts
private lastId = -5;
///
this.state = {
      scenes: [
        {
          title: `title ${this.lastId}`,
          position: this.lastId,
          id: (this.lastId++).toString()
        },
        {
          title: `title ${this.lastId}`,
          position: this.lastId,
          id: (this.lastId++).toString()
        },
        {
          title: `title ${this.lastId}`,
          position: this.lastId,
          id: (this.lastId++).toString()
        }
      ],
    }
```
Sphere objects work correctly
```tsx
{
  state.scenes.map(s => (
    <sphere
      key={s.id}
      name={s.title} diameter={1} segments={16} position={new Vector3(s.position, 0, 0)}>
    </sphere>
  ))
}
```
![image](https://user-images.githubusercontent.com/12235437/87359870-7994e980-c571-11ea-9b53-e07c8c337678.png)

Now, I want to create more spheres
```tsx
handleAdd() {
  this.setState(state => ({
    ...state,
    scenes: [
      ...state.scenes,
      {
        title: `title ${this.lastId}`,
        position: this.lastId,
        id: (this.lastId++).toString()
      }
    ]
  }));
}
///
<div>
  <button onClick={this.handleAdd}>Add</button>
</div>
{
  state.scenes.map(s =>
    <div key={s.id}>
      <button >{s.title}</button>
    </div>
  )
}
```
Spheres are created correctly
![image](https://user-images.githubusercontent.com/12235437/87360020-cc6ea100-c571-11ea-9dfb-5f98f5b24750.png)

And now I want to place titles of all spheres on full screen UI
```tsx
<adtFullscreenUi name="fullScreenUI">
  {
    state.scenes.map(s => (
      <textBlock top={s.position * 20} text={s.title} />
    ))
  }
</adtFullscreenUi>
```
But textBlocks are created only for initial three elements
![image](https://user-images.githubusercontent.com/12235437/87360121-1788b400-c572-11ea-9848-c6e16ea1af5f.png)

Working example is available here https://github.com/CAPCHIK/create-react-app-typescript-babylonjs

How I can use dynamic GUI objects?

**Comment by brianzinn** [comment]
I think you need to use a `key` attribute to avoid confusing the renderer (you should have a React warning about rendering lists in map without keys).  I have a working example of dynamic GUI for `<textBlock .../>` objects here:
https://github.com/brianzinn/react-babylonjs/blob/master/stories/babylonjs/GUI/with2DGUI.stories.js#L200

Thanks for the feedback and giving the repo a try.  If using `key` doesn't fix let me know and I can clone your repo and have a better look.

**Comment by CAPCHIK** [comment]
Of course, I forgot `key` property
But `key` can't resolve that problem, repo https://github.com/CAPCHIK/create-react-app-typescript-babylonjs already updated with `key` on `textBlock`

**Comment by brianzinn** [comment]
defo a bug in this library.  You can fix for now by adding a rectangle - like this:
```
<adtFullscreenUi name="fullScreenUI">
   <rectangle height='100%' width = '100%'>
    {
        state.scenes.map(s => (
           <textBlock key={s.id} top={s.position * 20} text={s.title} />
         ))
    }
  </rectangle>
</adtFullscreenUi>
```

I need to look at it further, but I think my renderer full screen UI component is not listening in properly on adding/removing GUI components.  Thanks for reporting.

**Comment by CAPCHIK** [comment]
On branch [_workaround_](https://github.com/CAPCHIK/create-react-app-typescript-babylonjs/tree/workaround) I create buffer on invisible textBlocks which easily transform to correct state (shown on title changing)

**Comment by brianzinn** [comment]
ok - thanks for the additional insight.  good to see the `isVisible` property being set correctly in state updates, but the underlying issue is that any control inheriting from 2D GUI class `container` (including the full screen container itself) needs to call `addControl(...)` / `removeControl(...)` dynamically as you add/remove GUI components to your scene.

**Comment by brianzinn** [comment]
Hi @CAPCHIK - thanks for using this library and sharing your issue with a reproduction.  I was able to fix finally - it was a combination of issues.  Mainly that these Full Screen UI and Advanced Dynamic Textures are not 2D controls, but just textures that support `addControl(...)` and `removeControl(...)`.  I am able to add and remove controls using test code like below:

```jsx
export const GUIFullScreen = () => {
  const [list, setList] = useState(['a', 'b']);
  const [adding, setAdding] = useState(true);
  console.log('rendering list:', list);
  return (
    <div style={{flex: 1, display: 'flex'}}>
      <button onClick={() => {
          if (adding) {
            setList([...list, 'c'])
          } else {
            setList(list.slice(0, list.length - 1));
          }
          setAdding(x => !x);
        }
      } />
      <WithGUI list={list}/>
    </div>
  )
}
```

I will get an NPM update out with the fix this week.  Have a nice weekend.

**Comment by brianzinn** [comment]
should be fixed in 2.2.3 - please reopen if you have any issues.
